docs/TPPDialect.md

```diff
@@ -0,0 +1,351 @@
+
+## Tensor Unpack
+The tensor operation `tensor.unpack` does a "block transpose" (n,m <-> m,n) copies.
+
+## VNNI Pack
+Packs into VNNI shape.
+Output type must have rank+1, last dim 2 or 4 depending on element type.
+Always out-of-place.
+```mlir
+  %1 = memref.alloc : memref<MxNxmxTy>
+  tpp.pack_vnni ins(%0) outs(%1) : (memref<64x32xbf16>, memref<32x32x2xbf16>) // TRANSFORM_NORM_TO_VNNI2
+  tpp.pack_vnni ins(%0) outs(%1) : (memref<64x32xf8>, memref<16x32x4xf8>) // TRANSFORM_NORM_TO_VNNI4
+  tpp.pack_vnni ins(%0) outs(%1) : (memref<16x32x4xf8>, memref<16x32x4xf8>) // Already packed
+```
+If the type is the same, this is just a notification that the shape is VNNI packed.
+It should be fused, but can be lowered as `COPY` with `TRANSFORM_NORM_TO_VNNI2` flag.
+
+## VNNI Unpack
+We don't need VNNI unpack, since we only use it for matmul and the result is unpacked.
+but if we do, same observations as above.
+
+## Arith
+Perform unary arithmetic operation.
+Types must be the same.
+```mlir
+  tpp.square ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // X2 op
+  tpp.sqrt ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // SQRT op
+  tpp.negate ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // NEGATE op
+  tpp.reciprocal ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // RECIPROCAL op
+  tpp.exp ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // EXP op
+```
+
+## Activation
+Non-linear activation functions.
+Types must be the same.
+```mlir
+  tpp.relu ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // RELU op
+  tpp.tanh ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // TANH op
+  tpp.sigmoid ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // SIGMOID op
+  tpp.gelu ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // GELU op
+  tpp.leaky_relu ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // LEAKY_RELU op
+  tpp.elu ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // ELU op
+```
+
+## Quant/Dequant
+Ignore for now.
+Should be as simple as the other unary ones.
+
+## Scatter/Gather
+Ignore for now.
+Should be as simple as the other unary ones.
+
+# Binary Ops
+
+## Element-wise Arith
+Apply element-wise ops..
+Operands of the same type, broadcast by operator.
+If operand's producer is `tpp.broadcast`, add appropriate `BCAST_*` flag to op.
+
+Generic out-of-place syntax:
+```mlir
+  %2 = memref.alloc : memref<NxMxTy>
+  tpp.add ins(%0, %1) outs(%2) : (memref<NxMxTy>, memref<NxMxTy>, memref<NxMxTy>) -> memref<NxMxTy> // ADD
+  tpp.sub ins(%0, %1) outs(%2) : (memref<NxMxTy>, memref<NxMxTy>, memref<NxMxTy>) -> memref<NxMxTy> // SUB
+  tpp.mul ins(%0, %1) outs(%2) : (memref<NxMxTy>, memref<NxMxTy>, memref<NxMxTy>) -> memref<NxMxTy> // MUL
+  tpp.div ins(%0, %1) outs(%2) : (memref<NxMxTy>, memref<NxMxTy>, memref<NxMxTy>) -> memref<NxMxTy> // DIV
+  tpp.muladd ins(%0, %1) outs(%2) : (memref<NxMxTy>, memref<NxMxTy>, memref<NxMxTy>) -> memref<NxMxTy> // MULADD
```

Comment:
In libxsmm TPP we support muladd as binary because the output *must* be aliasing with one of the inputs (the additive one). In principle the muladd should be ternary in the generic fashion: d = a * b + c (whereas we really implement only c = a * b +c in libxsmm TPP)